rss experiment input complete

ajCameron opened the pull request:
Output to follow - but the input side is currently functional so seemed a good idea to get it merged in and check the linters are still passing e.t.c.

ajCameron committed:
Preliminary work towards RSSIO

 - requirements.txt modified with the chosen feed library
 - RSSIO and Input classes defined and partially written
 - Very trivial example - reading from The Guardian - defined and running
 - bones of the process logic defined, but not working

ajCameron committed:
RSS input complete

 - polling interval set so the sites are evenly distributed in the polling time
 - basic example added - polling theguardian/word and engadget.com
 - input event parsing currently done in the main class
 - live updating sites has not been tested

ajCameron committed:
Updated examples/rss_input.py

Linting on windows continues problematic.

 - Unused objects removed
 - measures taken to log all input events - event unexpected ones

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+
+import asyncio
+from typing import Optional, Set, Sequence, Type, Any, Union, List, SupportsInt
+
```
This is a personal preference thing, but I like to keep typing separate from the other imports

```suggestion
from typing import Optional, Set, Sequence, Type, Any, Union, List, SupportsInt

import asyncio
```

ajCameron replied:
Updated to seperate

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+    def __init__(
+        self,
+        *args: Optional[Any],
+        polling_every: Union[str, bytes, SupportsInt],
```
Couldn't this be a `@property`?
(The long term goal is to infer setup information form the property objects, as reflecting on the arguments to `__init__` is tricky)

ajCameron replied:
It can be, and should.

Have added it to style guide notes (as I think this is generally preferable?)

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+    def __init__(
+        self, sites: List[str], polling_every: int, startup_queue_depth: int = 5
+    ) -> None:
+        super(Input, self).__init__()  # pylint: disable=bad-super-call
```
```suggestion
        super().__init__()
```
newer python is smarter for single-inheritance; the pylint warning is poorly worded in explaining what it wants you to do

ajCameron replied:
ohh - yes. I had totally misunderstood. Corrected.

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+            self._polling_every,
+        )
+
+        for target_site in self._sites_iter:
+
+            if target_site not in self.sites_started:
+                asyncio.get_running_loop().create_task(self.startup_site_feed(target_site))
+                # Delay before getting the next site
+                # stagger site reads to prevent overwealming anything
+                await asyncio.sleep(self.polling_every)
+                continue
+
+            asyncio.get_running_loop().create_task(self.poll_site_feed(target_site))
+            # Delay before getting the next site
+            # stagger site reads to prevent overwealming anything
+            await asyncio.sleep(self.polling_every)
```
I'm not sure if it improves readability, but maybe

```suggestion
        for target_site in self._sites_iter:

            if target_site not in self.sites_started:
                future = self.startup_site_feed(target_site)
            else:
                future = self.poll_site_feed(target_site)

            asyncio.get_running_loop().create_task(future)
            # Delay before getting the next site
            # stagger site reads to prevent overwealming anything
            await asyncio.sleep(self.polling_every)
```

javajawa replied:
`asyncio.get_running_loop()` could even be determined at `__init__` time, to be honest

exachixkitsune replied:
I meant to say "I agree with this, it removes an implicit branch and also avoids a weird branch to look at in the code" but SOMEONE misread "Commit Suggestion" as "Comment on suggestion"
(it was me, I misread github)

exachixkitsune replied:
ANYWAY
line 255 here has an errant ). Any change would want to remove it

javajawa replied:
I sneakily fixed the suggestion at some point :3

ajCameron replied:
> `asyncio.get_running_loop()` could even be determined at `__init__` time, to be honest

Loop isn't up at __init__ - but it can wrap it as a property - which may improve efficiency somewhat.

ajCameron replied:
Updated to improve readability

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+        self._logger.info("Reading from site %s", site_url)
+
+        # Read the site feed
+        site_newsfeed = feedparser.parse(site_url)
```
Is feedparser using asyncio? If not, can we do the HTTP connection with aiohttp and then give it the content, so there isn't a blocking call here?

ajCameron replied:
Now updated to do so.

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+        # THUS ONLY THE COMBINATION OF GUID AND SITE_URL SHOULD BE ASSUMED UNIQUE
+        # This combination is used to determine if events have already been sent
+
+        input_rss_event = RSSInputEvent(
```
Potential improvement, again don't know if it would actually improve readability: take this and all the functions called creating this object, and put them in a Builder/Marshler object -- essentially this class deals with wire transit (or Extract/Load parts of ETL) and have the transformation be distinct code.

(Also means you could possibly do weird plug-and-play stuff with having different marshlers in the future)

ajCameron replied:
Having looked at it both ways, I think it definitely does - and allows for greater flesxibility.

Updated.

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+            )
+
+            # This is not a fantastic guid - but one needs to be set for internal tracking
+            entry_guid = f"{site_url}:{entry_title}"
```
Minor: I'd be more included to go with link and date

ajCameron replied:
That does, indeed, make more sense. Corrected.

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+        self, entry: feedparser.util.FeedParserDict, site_url: str, problem_list: List[str]
+    ) -> str:
+
+        entry_title = self._extract_title(entry=entry, problem_list=problem_list)
```
This is only used in the `AttributeError` block, possible doesn't want to be up here

javajawa commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,514 @@
+        self._logger.info("%s entries from %s transmitted", transmitted_count, site_url)
+
+    @staticmethod
+    def _extract_title(entry: feedparser.util.FeedParserDict, problem_list: List[str]) -> str:
```
These field extractors appear to have a lot of common code. Would something similar to the `Dict.get()` semantics be easier to work with, where the user supplies the default in? The additions to `problem_list` could be based off a list of required fields.

Alterantively
```python
REQUIRED_FIELDS = {
  "title": "TITLE_NOT_SET",
  ...
}

def _extract_field(entry, field: str, problem_list: List[str]) -> Optional[str]:
  if has_attr(entry, field):
    return get_attr(entry. field)
  if field not in REUQIRED_FIELDS:
    return None
    
  problem_list.append(f"entry did not contain {field}")
  return REQUIRED_FIELDS[field]
```

ajCameron replied:
So - my thinking for this.

I considered a generic method, however some poking around revealed that there are lot of different ways which people have implemented the RSS standard. The current methods are more placeholders, with the aim to replace them with more sophisticated parsing methods when I have time to prod some of the weirder edge cases with a stick.

This might, or might not, be worth actually doing. Will give some more thought to it.

ajCameron replied:
Decided it was not worth the additonal LOC - if anyone needs to do custom extractors, they can add them later.

Kept the guid and pubDate methods, as those 
 - for guid, we need _something_ - so included a fallback method
 -  If the pubDate field is not filled then the source is producing very malformed RSS - which seemed worthy of a call to the log

javajawa reviewed (commented):
Code looks solid; got a few questions and ideas for simplifying things.

I'd like to sequence this after #37 and #38 which changes some of the underlying APIs; I think the best thing to do will be to rebase this branch once those are merged to keep a linear history, and also remove some of the linting commits that have no semantic value.

exachixkitsune commented on requirements.txt:
```diff
@@ -8,4 +8,6 @@ py-cord~=1.7
\ No newline at end of file
+win10toast~=0.9; sys_platform == "win32"
+
+feedparser~=6.0.8
```
```suggestion
feedparser~=6.0.8

```
Just adding a newline here in requirements.txt to avoid it thinking something changed with the prev line when nothing changed.

ajCameron replied:
Very sensible - adopted

exachixkitsune commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,512 @@
+    _output: None
+
+    _sites: List[str]  # A list of sites to poll for RSS update events
+    _polling_every: int  # Total re-poll time for the sites in the _sites list
```
Presuming seconds?

ajCameron replied:
Indeed - added.

exachixkitsune commented on src/mewbot/io/rss.py:
```diff
@@ -0,0 +1,512 @@
+        self._sites = sites
+        self._sites_iter = cycle(iter(self._sites))
+
+        self._polling_every = polling_every  # The time taken to poll ever site in sites once
```
Do we want to warn if this polling interval is very small? Not to enforce a min, but to let the user know if they've entered a bad value.

ajCameron replied:
This makes sense - and seems part of making the project more user friendly in general. Will add.

exachixkitsune commented:
Erroneous commit removed via reset
Errant branch is https://github.com/mewler/mewbot/tree/rss_experiment_errantcommit if anything major was lost.

exachixkitsune reviewed (commented):
Minor comments only; the other review covered everything major.

exachixkitsune reviewed (dismissed):
I am content from my point of view; the other comments remain.

ajCameron committed:
Code enhancements suggested by review - (1/n)

 - reordering imports to isolate typing - in line with project overall style
 - Moved polling_every over to a property - to assist with introspecting to determine object requirements
 - Corrected malformed super, allowing for disabling a pylint exception
 - improved readability in site for loop
 - moved to using asyncio for http request before feed parsing

ajCameron committed:
Code enhancements suggested by review - (2/n)

 - Moved the marshalling/parsing logic to its own class
 - Now calculating fallback guid with pubdate, instead of entry title - which should prove more reliable
 - Optimizing fallback entry calculation
 - Added newline to the end of requirements.txt to prevent tripping change detection
 - Added clarification that polling interval is in seconds
 - Added warning if polling interval is less than a short_warning_interval
 - tidying up comments

ajCameron committed:
Code enhancements suggested by review - (3/n)

 - Moved internal state of RSSInput to a dataclass to cut down on variables in the main class

ajCameron committed:
Code enhancements suggested by review - (3/n)

 - Moved internal state of RSSInput to a dataclass to cut down on variables in the main class
 - Also allowed for more efficient checking, and removed a (small) memory leak for long running processes

ajCameron committed:
Code enhancements suggested by review - (4/n)

 - Added type information to RSSInputState:start
 - Very weirdly, mypy is hanging - instead of returning - which, glancing at it, I thought was a successful run. Need to check why.

ajCameron committed:
Updated example

rss_input.yaml updated
 - To the new standard
 - Tested, working

ajCameron committed:
DRYing out the code base by removing individual field extractors

Logic was that parsing these fields may require custom methods in the future - so adding them as placeholders
 - Individual extraction methods remove and replaced with generic
 - method with fallback retained for guid
 - method with more comprehensive logging retained for pubDate

ajCameron committed:
Added tests to io.rss

- Basic tests
- That the IOConfig can load from yaml
- That some protected attributes can/cannot be set

ajCameron committed:
Grouped inputs

- Notably, this was not an issue raised during a run of tools/lint.py

exachixkitsune reviewed (dismissed):
Seems fine to me.

javajawa reviewed (dismissed):
Looks good asides from test coverage (and maybe cleaning up the giant commit message of doom)

ajCameron committed:
Experimental adding a single test

 - trying to debug a problem in a different branch - seeing if this is the thing which breaks all the linters

ajCameron committed:
Adding additional tests to meet minimal testing requirement

 - Should have hit minimum testing requirement for io.rss
 - Not _well_ mind - that comes later

ajCameron committed:
Adding tests to boost coverage

 - Properties tested
 - Running the input tested

ajCameron commented:
Issues with test coverage are resolved. Hopefully just one more round of approvals and can get it on mainline.
